Binding universe constraints in Definition/Inductive/etc...

mattam82 opened the pull request:
The commit message mostly says what this is. As discussed at the last WG and requested a few times, we lacked a way to give constraints directly at the definition level, this is now made possible, using syntax "@{i j k l | i < j, l <= k}" . 
For universes and constraints respectively, there is an additional boolean switch to indicate if e.g. at the end of the definition or proof additional universes and/or constraints can remain (I chose "+", I'm not sure notations based on "." like "..."  can be made to work robustly, "?" might be another option, "_" would be confusing I think). For compatibility @{} and @{u v} allow extensibility of constraints only, and @{| } has to be used to mean no undeclared constraint (note the space, it's another parsing issue, conflicting with the record declaration keywords). This is quite asymmetrical and ugly, we could also make the choice to break compatibility and force users to write @{u v|+} instead (probably just a sed script would be enough to fix scripts, but that's pretty pervasive source incompatibility).
 
Allow declaring universe constraints at definition level.

Introduce a "+" modifier for universe and constraint declarations to
indicate that these can be extended in the final definition/proof.  By
default [Definition f] is equivalent to [Definition f@{+|+}], i.e
universes can be introduced and constraints as well. For [f@{}] or
[f@{i j}], the constraints can be extended, no universe introduced, to
maintain compatibility with existing developments. Use [f@{i j | }] to
indicate that no constraint (nor universe) can be introduced. These
kind of definitions could benefit from asynchronous processing.

Declarations of universe binders and constraints also works for
monomorphic definitions. TODO: register the global binders named by the
user accordingly. Needs a bit more work to make it correct on proofs.

JasonGross commented:
Would it be worth it to have `Set Suggest Universe Annotations ` a la `Set Suggest Proof Using `?  Then I could make a python porting script based on this if you ever decide you want to flip the meaning of `@{i j}` or `@{}`.

SkySkimmer commented:
>@{| } has to be used to mean no undeclared constraint (note the space, it's another parsing issue, conflicting with the record declaration keywords)

Is there a way to special case that?

If possible I would like _ to mean exactly one more universe (possibly repeated, eg @{_ _} for 2 universes). I don't think constraints are really suited to this counting since they're transitive though.
I tried implementing that (https://github.com/SkySkimmer/coq/commit/fb3ff4513391c72106f6c9c29963e46ff78e180a) but I found that `Definition foo@{_ i} (A:Type@{i}) : Type := A.` would reorder the binder so `i` would be first and I couldn't figure out how to avoid it.

mattam82 commented:
@JasonGross that's a possible extension yes

mattam82 commented:
@SkySkimmer Maybe a camlp5 guru can do that. 

About the ordering indeed your feature requires a bit more work I think... you could change the code so that at interpretation time _ i is turned into Top.foo i and do the universe_context linearisation according to that information I suppose, no?

SkySkimmer commented:
>the universe_context linearisation

I don't know what that is.

maximedenes commented:
@mattam82 do we want this in 8.7?

maximedenes commented:
@mattam82 ping what is the target milestone? 8.7 or 8.8?

mattam82 commented:
I wanted this in 8.7, sorry for the delay.

mattam82 commented:
Given that we didn't discuss it at the WG and it needs more work, I'm fine with postponing to 8.8

mattam82 commented:
@SkySkimmer we do typechecking with a _set_ of universes (associating user-given names to universes) but when declaring definitions we linearize it to an array of universes. That's where the user-given information is used again to put the universes the user gave in the right order in the array. The problem here is that we have no information on the "_" universe, so we don't know what position it should have in this array. Is that clearer?

SkySkimmer commented:
@mattam82 
>when declaring definitions we linearize it to an array of universes

If you mean the `UState.universe_context` and wrappers the problem is that the array passes through https://github.com/coq/coq/blob/trunk/kernel/term_typing.ml#L276 where it is de-linearised back into a set, union'd with some side effect set then linearised at https://github.com/coq/coq/blob/trunk/kernel/term_typing.ml#L282 according to `Univ.Level.compare` ie `foo.n < foo.m` when `n < m`.
So the API pretends to respect the user given information but actually uses the fact that when doing `Definition foo@{i j} ...` the universe `i` is bound to gets created before the one for `j`.

mattam82 commented:
Ah, good catch, I hadn't realized that. How to fix it is another issue...

SkySkimmer commented:
It doesn't seem to detect constraints `u <= Set`:
``` coq
Set Universe Polymorphism.
Definition foo@{u| } := nat : Type@{u}.
About foo.
foo@{u} : Type@{u}
(* u |= Set <= u
         *)
```

JasonGross commented:
I believe it is a limitation of the current system that user-given polymorphic universes cannot be instantiated with `Prop`, so that constraint is vacuous.

SkySkimmer commented:
See mattam82/coq#7 for `@{|}`.

SkySkimmer commented on vernac/command.ml:
```diff
@@ -219,8 +219,9 @@ let do_definition ident k pl bl red_option c ctypopt hook =
-          ident ~term:c cty ctx ?pl ~implicits:imps ~kind:k ~hook obls)
+      let pl = Option.map (fun d -> d.univdecl_instance) pl in
+      ignore(Obligations.add_definition
+          ident ~term:c cty ctx ?pl:pl ~implicits:imps ~kind:k ~hook obls)
```
`?pl:pl` can just be `?pl`

mattam82 commented:
Indeed u >= Set for any polymorphically bound universe, and u > Set for any global universe.

mattam82 commented:
Rebased on master, cleaned up and now integrating 215a94e for parsing foo@{} to check that foo has no universe bindings.

silene commented:
For the record, I don't feel `@{i j}` meaning extensible constraints is that ugly. My guess is that most of the definitions in practice induce constraints, so the case `@{i j|+}` will be much more common than `@{i j|}` and thus is worth abbreviating into `@{i j}`, which also preserves backward compatibility..

mattam82 commented:
Yes, I realized that too, it's a fine abbreviation.

maximedenes commented:
@mattam the test suite seems to fail

mattam82 commented:
Rebased and fixed. I cleaned-up Proof_global.close_proof a bit and added comments, we will improve on that ones the patches by @ppedrot on de Bruijn contexts arrive at this level.

SkySkimmer commented on vernac/command.ml:
```diff
@@ -1018,6 +1018,7 @@ let build_wellfounded (recname,pl,n,bl,arityc,body) poly r measure notation =
   let binders_rel = nf_evar_context !evdref binders_rel in
   let binders = nf_evar_context !evdref binders in
   let top_arity = Evarutil.nf_evar !evdref top_arity in
+  let pl = Option.map (fun d -> d.univdecl_instance) pl in
```
In the extensible case do we not get false errors at line 1029 below?

SkySkimmer commented:
I made mattam82/coq#8 such that for instance
``` coq
Definition foo@{u +} := Type -> Type@{v} -> Type@{u}.
(* foo@{u Top.8 v} = 
Type@{Top.8} -> Type@{v} -> Type@{u}
     : Type@{max(u+1, Top.8+1, v+1)}
*)
```
(also works without the `@...` but then you get `foo@{Top.xx v u}` since it appears last in the term.)

mattam82 commented:
Ok, I'll rebase

Zimmi48 commented:
This PR passed the test. Is it ready to merge?

mattam82 commented:
Yes it is.
